Egglaying

AlecWhite opened the pull request:
This PR makes monster lay eggs:

- Added `EGGLAYING` flag to monster.
- Added `egglaid` effect of 2 days max duration.
- Monsters with the `EGGLAYING` flag will lay an egg every 2 days.
- Chickens now lay eggs.
- Added `òn_idle()` method and called it in `move()`.
- `òn_idle()` method defaults to `stumble()` when the monster can't do any special action.
- `EGGLAYING` monsters will only lay eggs on furniture with the `NEST` flag.
- Added `lay_egg` attack and add it to chickens.
- Added animal box and its recipe.
- Added `NEST` flag for furniture
- Animal box is used the same way as pet carrier but they are craftable.
- Added makeshift nest and its recipe. 


As always I tried to be as less invasive as possible, and due to my little knowledge of how things work I opted to do it the easiest way.

Will work on a poultry farm soon.

Coolthulhu commented:
I'd leave that for much later.
Egg laying will not work as nicely as milking - the result will be chickens laying fresh eggs when you pass by and random eggs littering the landscape.

For proper egg laying, the animals should have an AI that directs them into their nest before laying eggs and more "conscious" egg laying.
Not to mention that there is still no way to herd chickens properly.

BrettDong commented on data/json/effects.json:
```diff
@@ -1652,5 +1652,12 @@
+        "type": "effect_type",
+        "id": "laidegg",
+        "name": [ "Laid an egg" ],
+        "desc": [ "This creature has laid an egg recently" ],
```
There should be a period. Some translators complained to me about the inconsistency of the period symbols at the end of the strings in game recently so I'm paying attention to that in new PRs.

BorkBorkGoesTheCode commented:
Could some method of converting eggs into chickens be implemented now, or in a later PR?

AlecWhite commented:
>For proper egg laying, the animals should have an AI that directs them into their nest before laying eggs and more "conscious" egg laying.
Not to mention that there is still no way to herd chickens properly.

If that's the general consensus, I'm going to see if I can understand and work a way with the AI(is it basically in monmove.cpp for the AI and game::monmove() with some tidbits spread in monster.cpp to actually make them move right?) to make them move to a nest tile. The whole scanning an area from the chicken's position to search of a nest is going to the FUN to figure it out.
I think if I change the pet menu stuff and make the examinemonsters.cpp/.h I could make so chickens only lay eggs if fed and after the laidegg effect is off.
But again, don't take that as a promise, I'm too noob at C++ to work fast or work well. And generally documentation regarding the AI is too little to compensate for my noobness, so it is going to be a try and error thing(pretty much like everything else I made so far). But then it will be fun to try.

>Could some method of converting eggs into chickens be implemented now, or in a later PR?

Short answer yes, long answer:
Wouldn't be too hard to make roosters go to a mob and mate with them(I think), but I still have to see a way to make roosters *only* mate with chickens and not with any other entity.
But then I would need a way to track the health of the fertilized eggs, and make the eggs check for the temperature each turn and then make them hatch. And another way to make chicks grow.

kevingranade commented:
You're going to want two different pieces of code for placing eggs.  The main piece would retroactively place eggs when egg-laying birds are loaded onto the map, the secondary piece would have the same birds move to a nesting area when they are ready to lay eggs, then produce eggs.  A key thing to remember is which is primary and which is secondary.  We expect the vast majority of eggs to be placed when the player is not nearby, so the "on-load" code places most of the eggs.  The AI code is just there to keep things consistent in the rare case that the player happens to be nearby when birds are laying eggs.

Also you're going to want some special handling for eggs disappearing, to avoid the player coming across huge caches of eggs late in the game.  In reality it's cyclic and you would expect to find similar amounts of eggs independent of date (unless you add in something for seasonal laying rates).

Night-Pryanik commented on src/monstergenerator.cpp:
```diff
@@ -446,6 +446,7 @@ void MonsterGenerator::init_flags()
     flag_map["PATH_AVOID_DANGER_2"] = MF_AVOID_DANGER_2;
     flag_map["PRIORITIZE_TARGETS"] = MF_PRIORITIZE_TARGETS;
     flag_map["NOT_HALLUCINATION"] = MF_NOT_HALLU;
+	flag_map["EGGLAYING"] = MF_EGGLAYING;
```
Tab instead of spaces.

Night-Pryanik commented on src/mtype.h:
```diff
@@ -157,6 +157,7 @@ enum m_flag : int {
     MF_AVOID_DANGER_2,      // This monster will path around most dangers instead of through them.
     MF_PRIORITIZE_TARGETS,  // This monster will prioritize targets depending on their danger levels
     MF_NOT_HALLU,           // Monsters that will NOT appear when player's producing hallucinations
+	MF_EGGLAYING,			// This monster will lay eggs.
```
Tab instead of spaces.

AlecWhite commented:
Close as I would need stuff in https://github.com/CleverRaven/Cataclysm-DDA/pull/21272 to start taming the chickens so they don't flee from you.

kevingranade commented:
Even when tame, chickens scatter when people approach.  If you want to farm
chickens you either let them free range and go hunt down eggs, or you put
them in an enclosure.

DangerNoodle commented:
> Even when tame, chickens scatter when people approach. If you want to farm chickens you either let them free range and go hunt down eggs, or you put them in an enclosure.

Enclosing livestock is something players will likely do anyway, as you would expect tame animals to potentially wander off even if they do not actively flee the player, in addition to players wanting to prevent them from being harmed.

AlecWhite commented:
>Even when tame, chickens scatter when people approach. If you want to farm chickens you either let them free range and go hunt down eggs, or you put them in an enclosure.

My idea was doing something like http://dwarffortresswiki.org/index.php/DF2014:Nest_box coupled with https://www.omlet.co.uk/guide/chickens/chicken_care/how_to_tame_your_chickens

And for chickens laying eggs retroactively on on load, I would need a some kind of next, to control the amount of eggs laid, and to have a place for the eggs to go. Plus that would require to assign a nest for each chicken if you want them to lay egg.

kevingranade commented:
Something like a nest box is fine, it would just be a piece of furniture that's an ideal nest, so whatever nest-selection code we use would prefer it.

You don't get tame chickens unless you model raising them from chicks, the default is semi-tame chickens, so you need to handle that.  You can add hand-rearing if you like, but it will never be the default.

Yes you need code that finds a nest, you need code that regulates number if eggs laud, you need code to avoid crowding.  These are just what's required for the thing you're adding.

AlecWhite commented:
You convinced me to do it without monexamine stuff.

mouseturd commented:
Not sure if this is doable - but with my chickens at home, the only way to "herd" them is to chase them, pick them up, then throw them in their fenced enclosure.

Is there a way of "picking up" or "grabbing" a chicken or other small furry animal at this time?

kevingranade commented:
> Is there a way of "picking up" or "grabbing" a chicken or other small furry animal at this time?

There is not, but it wouldn't be that hard to add.

DangerNoodle commented:
> There is not, but it wouldn't be that hard to add.

Pet carrier?

mouseturd commented:
If grabber zombies can move me around the map, it would be nice if we
could do the same to animals.

On Jun 23, 2017 12:59 PM, "DangerNoodle" <notifications@github.com> wrote:

> There is not, but it wouldn't be that hard to add.
>
> Pet carrier?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/CleverRaven/Cataclysm-DDA/pull/21261#issuecomment-310719275>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQjUcF0aCGiOaQHP5HqYErW4PCrb_KLHks5sG-7qgaJpZM4N9QFL>
> .
>

Coolthulhu commented:
Don't implement it as an attitude, this will mess up the AI for other creatures. And also be slow, since attitude checks are common.
Instead, check the egg laying where you currently check the new attitude.
Or better yet: create an "on idle" method, which you invoke in that place, then do the egg laying check in that method.

AlecWhite commented:
What do you mean by "on idle" method? Calling a method that checks if the monster is fed, searches for a nest and returns a tripoint assigning it to `destination` at the end of `move()`'s destination logic at line 550?

I don't fully understand how the AI works or how `move()` works but I don't mind trying it any other way as long as I'm not too noob to be able to. So any help and ideas I can get would awesome. And if I'm doing this I would rather do it the best possible way I can.

Coolthulhu commented:
> What do you mean by "on idle" method?

Just wrapping the code in a new method, as in `monster::on_idle`. Not necessary, it's just that it would be cleaner than dumping it directly into `move`.

AlecWhite commented:
>Not necessary, it's just that it would be cleaner than dumping it directly into move.

Thanks, I understand now, and it makes sense after reading the cooment about `// TODO: Move this to attack_at/move_to/etc. functions`.

And without the attitude check in which part inside of `move()` would be the correct place to call it?

AlecWhite commented:
Time to see how to revert a branch in which I have merged the latest change of master. If anything github related weird things happens, it will be due to that.

AlecWhite commented:
Note to self: If you delete the last commit in a PR, it will auto-close.

Coolthulhu commented:
Now that I think about it, it would be better to put `on_idle` near the end of planning - that way egg layers don't path every turn, only on those turns they don't have a path.

AlecWhite commented:
Before the last `else` and assign `set_dest` the result of on_idle()?

AlecWhite commented:
Also I don't know what's the problem with gorgon now.

AlecWhite commented:
It seems to work fine moving on_idle() to monster::plan() if there isn't anything the monster needs to fight or flight.

I should probably now see how to make a better makeshift nest recipe which I could use some help with. 
And I don't know if I should make another separately PR for catching wild chickens/turkeys with the same pet carrier mechanics and add another craftable alternative for the pet carrier.

Coolthulhu commented on src/monmove.cpp:
```diff
@@ -534,6 +538,11 @@ void monster::move()
+    if( has_flag( MF_EGGLAYING ) && !has_effect( effect_laid_egg ) ) {
+        destination = on_idle();
+        moved = true;
+    }
```
Remove this section, now it is handled by planning

Coolthulhu commented on src/monmove.cpp:
```diff
@@ -341,7 +342,10 @@ void monster::plan( const mfactions &factions )
             set_dest( g->u.pos() );
-        } else {
+        }
+        else if( has_flag( MF_EGGLAYING ) && has_effect( effect_laid_egg ) ) {
```
Those checks should be in `on_idle`. That is, `on_idle` should be called unconditionally, but should do nothing for most monsters.
Also, looks like you inverted the effect check.

AlecWhite replied:
If it's called unconditionally and should do nothing for most monsters how do you check which monster to apply it?

AlecWhite replied:
Never mind, I figured it out, I think.

Coolthulhu replied:
Just that the two checks should be in `on_idle`, not here, in the outside.
If you're having problems because of the `unset_dest` below, move it into `on_idle`.

So `on_idle` would check if the monster is an egg layer and then try to find a nest for it, but for all other monsters it would just unset destination. It would also unset destination for egg layers who don't manage to find a nest (or don't want to, because of egg timer).

Coolthulhu commented on src/game.cpp:
```diff
@@ -13886,3 +13888,21 @@ std::vector<npc *> game::allies()
+            g->m.add_item_or_charges( critter.pos(), egg );
+            critter.add_effect( effect_laid_egg, ( HOURS( 48 ) ) );
+        }
+    }
```
This would be better off as a monster attack, as in `monattack.cpp`. Monster "attacks" don't need to be actual attacks, they're just abilities monsters use by themselves.

Here you go through all monsters, which is slow and unnecessary.

AlecWhite replied:
>Monster "attacks" don't need to be actual attacks, they're just abilities monsters use by themselves.

Didn't know that.

And what if I move mon_lay_egg() to monster::move() and add a check for egglaying flag and effect?

Coolthulhu replied:
Attack is a better way to do it: cleaner, more standard.

Coolthulhu commented on src/game.cpp:
```diff
@@ -5915,6 +5915,8 @@ void game::monmove()
 
         m.creature_in_field( critter );
 
+        mattack::lay_egg( &critter );
```
Don't call it here, you need to bind the attack to the monster as it is the case for all other attacks.

Coolthulhu commented on src/game.h:
```diff
@@ -881,6 +881,7 @@ class game
         bool handle_action();
         bool try_get_right_click_action( action_id &act, const tripoint &mouse_target );
         bool try_get_left_click_action( action_id &act, const tripoint &mouse_target );
+        void mon_lay_egg();
```
No longer used

Coolthulhu commented on src/monattack.cpp:
```diff
@@ -4647,3 +4648,19 @@ bool mattack::dodge_check( monster *z, Creature *target ) {
+    monster &critter = *z;
+    const furn_t &furn = g->m.furn( critter.pos() ).obj();
+
+    if( !critter.is_dead() && critter.has_flag( MF_EGGLAYING ) && furn.has_flag( TFLAG_NEST ) &&
```
Skip the egglaying check here - a monster that has `lay_egg` attack must be egglaying.
Would be clearer if you renamed it to `MF_SEEKS_NEST` or something like that.

AlecWhite replied:
@Coolthulhu I don't have any idea what you mean by `MF_SEEKS_NEST`, and I can't find a way to make a `lay_egg` attack.

I barely understand how the whole attack thing works and there isn't much documentation about it.
From what I believe the `for( const auto &sp_type : type->special_attacks )` loop in move() method in monmove.cpp tries somehow to get attack type of that monster, and then call it.

So I made `layegg()` into a bool and added `"special_attacks": [ [ "LAY_EGG" ] ],` to mon_chicken, but the game throws me that is an invalid special_attacks.

Defined `flag_map["LAY_EGG"] = MF_LAY_EGG;` in monstergenerator.cpp and added `MF_LAY_EGG` in mtype.h.

What else do I have to do to create a new special attack?

Coolthulhu replied:
> I don't have any idea what you mean by MF_SEEKS_NEST, and I can't find a way to make a lay_egg attack.

Your `MF_EGGLAYING` flag controls both the nest seeking and the egg laying. But that is not the proper way to do it, as one part of this behavior is just seeking nests and the other is laying eggs.
Seeking nests should happen in planning phase because it determines what the monster wants to do next.
Laying eggs should be an attack because it is what the monster actually does.

> I barely understand how the whole attack thing works and there isn't much documentation about it.

You have attacks declared in `monattack.h`, implemented in `monattack.cpp`, "named" in `monstergenerator.cpp` (in `init_attack` - just create a new entry), then bound to monster's special attack list in monster's json.
For this to be possible, all monster attacks need the same signature: `bool attack_name_here( monster *z );`.

You did most of that, but didn't do the `monstergenerator.cpp` part yet. This is why you're getting the error: the attack exists, but the game can't "find" it.

AlecWhite replied:
Made `lay_egg()` mattack, and it works if I spawn a chicken on a nest, but now for some reason it seems that the chickens aren't interested in going to the nest. Gonna see why and how to fix it.

Coolthulhu commented on src/monattack.h:
```diff
@@ -89,6 +89,7 @@ void rifle              (monster *z, Creature *target); // Automated M4
 void frag               (monster *z, Creature *target); // Automated MGL
 void tankgun            (monster *z, Creature *target); // Tankbot primary.
 void flame              (monster *z, Creature *target);
+void lay_egg            ( monster *z );
```
Move it up to the attacks above. It needs to be `bool lay_egg( monster *z);` or it won't be bindable properly.

Coolthulhu commented on src/monmove.cpp:
```diff
@@ -1445,3 +1446,21 @@ int monster::turns_to_reach( int x, int y )
+        }
+        set_dest( nest_pos );
+    }
+    else {
```
Spacing/newlines should be `} else {`

Coolthulhu commented on src/monattack.cpp:
```diff
@@ -4649,17 +4649,19 @@ bool mattack::dodge_check( monster *z, Creature *target ) {
 
-    if( !critter.is_dead() && critter.has_flag( MF_EGGLAYING ) && furn.has_flag( TFLAG_NEST ) &&
-        !critter.has_effect( effect_laid_egg ) ) {
+    if( !critter.is_dead() && furn.has_flag( TFLAG_NEST ) ) {
```
You can skip the dead check - it is already checked before executing the attack function.
But you do need to check the laid egg effect - this one won't be checked automatically.

AlecWhite commented:
Adding `on_idle()` on `plan()` never gets called and `unset_dest` seems to be there more as a safeguard. 
So barring making a new attitude, I moved it to `move()` at the end of the method, and it substitutes `stumble()` which is the default called when the monsters aren't fighting/fleeing, and added an `else` at the end of `on_idle()` to call `stumble`.
So now monsters that are idle will either do the special attack/action for when idle or stumble as they always did.

AlecWhite commented:
`15:32:19 ERROR: Format error at line 8273
15:32:19 <   },{
15:32:19 >   },
15:32:19 
15:32:19 HINT: Canonical output for this block was:`

No idea how to fix that.

Coolthulhu commented:
> No idea how to fix that.

Lint

DangerNoodle commented:
You can use the linting tool here: http://dev.narc.ro/cataclysm/format.html

BevapDin commented on data/json/furniture.json:
```diff
@@ -2769,5 +2769,34 @@
+        "deconstruct": {
+            "items": [
+                { "item": "2x4", "count": 10 },
+                { "item": "rag", "count": [30, 33] },
```
Deconstruction this gives *much* more rags and 2x4s than constructing it needs. Same with rags and bashing (construction requires 10 rags, bashing give 15-20).

BevapDin commented on data/json/items/tools.json:
```diff
@@ -8270,5 +8270,34 @@
     "max_charges": 100
+  },
+  {
+    "result": "washboard",
```
Any reason why you moved a *recipe* into this file (which is for item definitions)? Why did you even touch that recipe at all?

AlecWhite replied:
I mistakenly deleted that recipe and I mistakenly re-insert it on the wrong json.

BevapDin commented on src/monmove.cpp:
```diff
@@ -1445,3 +1445,21 @@ int monster::turns_to_reach( int x, int y )
+        const int range = 15;
+        // Loop to find the closest furniture with NEST flag and set it as destination
+        for( tripoint &p : closest_tripoints_first( range, pos() ) ) {
+            if( sees( p ) && g->m.has_flag( TFLAG_NEST, p ) ) {
```
Here you check for the flag via `map::has_flag` (which for example also considers flags of the *terrain*). In `mattack::lay_egg`, you check via `furn_t::has_flag`. This is inconsistent.

BevapDin commented on src/monattack.cpp:
```diff
@@ -4647,3 +4648,20 @@ bool mattack::dodge_check( monster *z, Creature *target ) {
+    monster &critter = *z;
+    const furn_t &furn = g->m.furn( critter.pos() ).obj();
+    // If the monster is in a tile with the NEST flag, lay egg
+    if( furn.has_flag( TFLAG_NEST ) ) {
```
There is `map::has_flag_furn` when you want to check only the furniture for a flag. No need to get the `furn_t` here.

BevapDin commented on src/monattack.cpp:
```diff
@@ -4647,3 +4648,20 @@ bool mattack::dodge_check( monster *z, Creature *target ) {
+    if( furn.has_flag( TFLAG_NEST ) ) {
+        item egg( "egg_bird", 0, 1 );
+        g->m.add_item_or_charges( critter.pos(), egg );
+        critter.add_effect( effect_laid_egg, ( HOURS( 48 ) ) );
```
Parentheses around `HOUR()` are useless.

BevapDin commented on src/monmove.cpp:
```diff
@@ -1445,3 +1445,21 @@ int monster::turns_to_reach( int x, int y )
+        // Loop to find the closest furniture with NEST flag and set it as destination
+        for( tripoint &p : closest_tripoints_first( range, pos() ) ) {
+            if( sees( p ) && g->m.has_flag( TFLAG_NEST, p ) ) {
+                set_dest( p );
```
You should probably leave the loop after this via `break`. Currently the game would go on and check all the other tiles within range as well. The monster would essentially go to the furthest visible point.

BorkBorkGoesTheCode commented:
I wonder if this could be used to create bee swarms which lay honey inside hollow trees.

AlecWhite commented:
>I wonder if this could be used to create bee swarms which lay honey inside hollow trees.
Everything is hardcoded, and wasn't made thinking of modularity and even less moddability, but wouldn't be too hard to do the bee swarms, as another hardcoded PR. Although it would probably need a few more requiremenst because of balance/realism and all that stuff.

And after saying this I should look at how to resolve conflicts in github.

AMurkin commented:
Incorrect merge. JSON structure is broken.

BevapDin commented on data/json/effects.json:
```diff
@@ -1678,6 +1678,11 @@
+        "name": [ "Disabled" ],
+        "desc": [ "This creature has laid an egg" ],
+        "max_duration": 2400
+    }
```
`,` missing.

BevapDin commented on data/json/items/tools.json:
```diff
@@ -8274,6 +8274,21 @@
+    "color": "light_brown",
+    "properties": [ [ "monster_size_capacity", "SMALL" ] ],
+    "use_action": "CAPTURE_MONSTER_ACT",
+    "flags": [ "TRADER_AVOID" ]
```
`}` and `,` and `{` missing.

Coolthulhu commented on data/json/construction.json:
```diff
@@ -2371,5 +2371,23 @@
+            [
+            { "id": "HAMMER", "level": 2 }
+            ]
+            ],
```
Is this linted? It looks really bad.

AlecWhite replied:
The linting tool doesn't like the fabrication field, so this time I'm going to pass it through without that field, and then add it later.

Coolthulhu commented on data/json/effects.json:
```diff
@@ -1678,6 +1678,13 @@
     {
         "type": "effect_type",
+        "id": "laid_egg",
+        "name": [ "Disabled" ],
```
Needs a name

Coolthulhu commented on src/monster.h:
```diff
@@ -323,6 +323,12 @@ class monster : public Creature, public JsonSerializer, public JsonDeserializer
 
+        /*
+        * Decides what to do in case the monster is idle.
+        * If more idle actions are added change if loop for switch case.
```
Not a switch, but rather something like monster attack - a json-coded function wrapper.

Coolthulhu commented on src/monstergenerator.cpp:
```diff
@@ -337,6 +337,7 @@ void MonsterGenerator::init_attack()
     add_hardcoded_attack( "RANGED_PULL", mattack::ranged_pull );
     add_hardcoded_attack( "GRAB", mattack::grab );
     add_hardcoded_attack( "GRAB_DRAG", mattack::grab_drag );
+    add_hardcoded_attack("LAY_EGG", mattack::lay_egg);
```
Styling

kevingranade commented:
This needs to have some code that triggers when an EGGLAYING creature is loaded on the map to find a nearby next and populate it with some eggs.

AlecWhite commented:
Wouldn't that assume that the area was safe of monsters and the player to not trigger the fleeing state?
I mean, part of making chickens non-tameable was to force them to first, protect them, secon; leave them alone and third; balance, as it would make otherwise an already very good source of food.

And now that I'm reminded of this. I'm still not completely happy with the balance of the egg spawn, after playing it for some time, as our most "hardcore" players may find it a bit too easy.

Would it be good if I added some way like with the milking to configure the cool down between laying eggs?

kevingranade commented:
I'm not sure we're on the same page, as I said earlier, the vast majority of eggs would be laid when the player isn't present, there needs to be code to retroactively place those eggs when the player enters the area.

To put this another way, the way it works now is birds *wait* for the player to arrive before laying eggs, as soon as a bird spawns on the map, it will make a beeline for the nearest nest and  immediately lay an egg.  To see this in action, spawn a bunch of chickens and nests, leave the area, and return after several days have passed.  All the chickens will run to the nests in unison and lay their eggs all at once.  This is not an ok way for this to work.

AlecWhite commented:
Closing due to private reasons.
